File: contacts-api/src/Contacts.Application/Services/ContactService.cs
using Contacts.Application.DTOs;
using Contacts.Application.Interfaces;
using Contacts.Domain.Entities;
using Contacts.Domain.Interfaces;

namespace Contacts.Application.Services;

public class ContactService : IContactService
{
    private readonly IContactRepository _repository;

    public ContactService(IContactRepository repository)
    {
        _repository = repository;
    }

    public async Task<PagedResultDto<ContactResponseDto>> GetAllActiveAsync(int page = 1, int pageSize = 10)
    {
        var contacts = await _repository.GetAllActiveAsync();
        var total = await _repository.CountActiveAsync();

        var items = contacts
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(ContactResponseDto.FromEntity);

        return new PagedResultDto<ContactResponseDto>(items, total, page, pageSize);
    }

    public async Task<ContactResponseDto> GetByIdAsync(Guid id)
    {
        var contact = await _repository.GetActiveByIdAsync(id)
            ?? throw new KeyNotFoundException($"Contato {id} não encontrado ou inativo.");

        return ContactResponseDto.FromEntity(contact);
    }

    public async Task<ContactResponseDto> CreateAsync(ContactRequestDto dto)
    {
        var contact = Contact.Create(dto.Name, dto.BirthDate, dto.Gender);
        await _repository.AddAsync(contact);
        await _repository.SaveChangesAsync();
        return ContactResponseDto.FromEntity(contact);
    }

    public async Task<ContactResponseDto> UpdateAsync(Guid id, ContactRequestDto dto)
    {
        var contact = await _repository.GetActiveByIdAsync(id)
            ?? throw new KeyNotFoundException($"Contato {id} não encontrado ou inativo.");

        contact.Update(dto.Name, dto.BirthDate, dto.Gender);
        await _repository.UpdateAsync(contact);
        await _repository.SaveChangesAsync();
        return ContactResponseDto.FromEntity(contact);
    }

    public async Task DeactivateAsync(Guid id)
    {
        var contact = await _repository.GetByIdAsync(id)
            ?? throw new KeyNotFoundException($"Contato {id} não encontrado.");

        contact.Deactivate();
        await _repository.UpdateAsync(contact);
        await _repository.SaveChangesAsync();
    }

    public async Task DeleteAsync(Guid id)
    {
        var contact = await _repository.GetByIdAsync(id)
            ?? throw new KeyNotFoundException($"Contato {id} não encontrado.");

        await _repository.DeleteAsync(contact);
        await _repository.SaveChangesAsync();
    }
}

File: contacts-api-node/src/application/services/ContactService.ts
import { IContactService } from "../interfaces/IContactService";
import { CreateContactDto, UpdateContactDto, ContactResponseDto, PagedResultDto } from "../dtos/ContactDto";
import { Contact } from "../../domain/entities/Contact";
import { Gender } from "../../domain/enums/Gender";
import { DomainException } from "../../domain/exceptions/DomainException";

export class ContactService implements IContactService {
  private contacts: Contact[] = [];

  public async getAll(page: number = 1, pageSize: number = 10): Promise<PagedResultDto> {
    const activeContacts = this.contacts.filter(c => c.isActive);
    const totalCount = activeContacts.length;

    const items = activeContacts
      .slice((page - 1) * pageSize, page * pageSize)
      .map(this.toResponseDto);

    return {
      items,
      totalCount,
      page,
      pageSize
    };
  }

  public async getById(id: string): Promise<ContactResponseDto> {
    const contact = this.contacts.find(c => c.id === id && c.isActive);
    if (!contact) {
      throw new DomainException(`Contato ${id} não encontrado ou inativo.`);
    }
    return this.toResponseDto(contact);
  }

  public async create(dto: CreateContactDto): Promise<ContactResponseDto> {
    const birthDate = new Date(dto.birthDate);
    const contact = Contact.create(dto.name, birthDate, dto.gender);
    this.contacts.push(contact);
    return this.toResponseDto(contact);
  }

  public async update(id: string, dto: UpdateContactDto): Promise<ContactResponseDto> {
    const contact = this.contacts.find(c => c.id === id && c.isActive);
    if (!contact) {
      throw new DomainException(`Contato ${id} não encontrado ou inativo.`);
    }
    const birthDate = new Date(dto.birthDate);
    contact.update(dto.name, birthDate, dto.gender);
    return this.toResponseDto(contact);
  }

  public async deactivate(id: string): Promise<void> {
    const contact = this.contacts.find(c => c.id === id);
    if (!contact) {
      throw new DomainException(`Contato ${id} não encontrado.`);
    }
    contact.deactivate();
  }

  public async delete(id: string): Promise<void> {
    const index = this.contacts.findIndex(c => c.id === id);
    if (index === -1) {
      throw new DomainException(`Contato ${id} não encontrado.`);
    }
    this.contacts.splice(index, 1);
  }

  private toResponseDto(contact: Contact): ContactResponseDto {
    return {
      id: contact.id,
      name: contact.name,
      birthDate: contact.birthDate.toISOString(),
      age: contact.age,
      gender: contact.gender as Gender,
      isActive: contact.isActive,
      createdAt: contact.createdAt.toISOString(),
      updatedAt: contact.updatedAt?.toISOString() || null
    };
  }
}
